File: go/sekreto/providers.go
package sekreto

import (
	"encoding/json"
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"syscall"

	plugin "github.com/voxgig/plugin/go/plugin"
)

type Provider interface {
	Lookup(name string) (string, bool, error)
	Describe() string
}

type ProviderSpec struct {
	Kind           string            `json:"kind,omitempty"`
	Name           string            `json:"name,omitempty"`
	Prefix         string            `json:"prefix,omitempty"`
	File           string            `json:"file,omitempty"`
	Values         map[string]string `json:"values,omitempty"`
	Dir            string            `json:"dir,omitempty"`
	Addr           string            `json:"addr,omitempty"`
	Token          string            `json:"token,omitempty"`
	Mount          string            `json:"mount,omitempty"`
	KV             int               `json:"kv,omitempty"`
	VaultNamespace string            `json:"vaultnamespace,omitempty"`
	// Auth logs hashicorp in for a token instead of being handed one.
	Auth         *AuthSpec `json:"auth,omitempty"`
	Command      string    `json:"command,omitempty"`
	Namespace    string    `json:"namespace,omitempty"`
	Home         string    `json:"home,omitempty"`
	Profile      string    `json:"profile,omitempty"`
	Backend      string    `json:"backend,omitempty"`
	Reason       string    `json:"reason,omitempty"`
	Region       string    `json:"region,omitempty"`
	KeyID        string    `json:"keyid,omitempty"`
	Secret       string    `json:"secret,omitempty"`
	Session      string    `json:"session,omitempty"`
	Project      string    `json:"project,omitempty"`
	Vault        string    `json:"vault,omitempty"`
	Tenant       string    `json:"tenant,omitempty"`
	ClientID     string    `json:"clientid,omitempty"`
	ClientSecret string    `json:"clientsecret,omitempty"`
	// azure: where to log in / where IMDS answers. gcp: where the
	// metadata server answers. Overridable for tests and for clouds with
	// nonstandard endpoints.
	LoginAddr    string `json:"loginaddr,omitempty"`
	ImdsAddr     string `json:"imdsaddr,omitempty"`
	MetadataAddr string `json:"metadataaddr,omitempty"`
	ApiVersion   string `json:"apiversion,omitempty"`
	Config       string `json:"config,omitempty"`
	Environment  string `json:"environment,omitempty"`
	Path         string `json:"path,omitempty"`
	Passphrase   string `json:"passphrase,omitempty"`
	// VaultKey is which key in a minivault file to open with, defaulting
	// to `master`. Named apart from Key and KeyID because those already
	// mean a secret name and an AWS access key id.
	VaultKey   string `json:"vaultkey,omitempty"`
	Iterations int    `json:"iterations,omitempty"`
	Create     bool   `json:"create,omitempty"`

	Provider Provider `json:"-"`
}

// AuthSpec is how the hashicorp provider logs in for a token instead of
// being handed one.
type AuthSpec struct {
	Method   string `json:"method,omitempty"`
	Mount    string `json:"mount,omitempty"`
	Role     string `json:"role,omitempty"`
	Jwt      string `json:"jwt,omitempty"`
	JwtFile  string `json:"jwtfile,omitempty"`
	RoleID   string `json:"roleid,omitempty"`
	SecretID string `json:"secretid,omitempty"`
}

type EnvProvider struct {
	Prefix string
	Source map[string]string
}

func (provider *EnvProvider) Lookup(name string) (string, bool, error) {
	key, err := EnvKey(name, provider.Prefix)
	if nil != err {
		return "", false, err
	}

	if nil != provider.Source {
		value, has := provider.Source[key]
		return value, has, nil
	}

	value, has := os.LookupEnv(key)
	return value, has, nil
}

func (provider *EnvProvider) Describe() string {
	if "" != provider.Prefix {
		return "env:" + provider.Prefix
	}
	return "env"
}

type DotenvProvider struct {
	File   string
	Prefix string
	// Guards the memoised state below: a Sekreto may resolve from several
	// goroutines, and a racy map read is either a crash or a zero value - a
	// MISS where the store does hold the secret, which falls through to a
	// weaker store.
	mu     sync.Mutex
	values map[string]string
}

func (provider *DotenvProvider) load() (map[string]string, error) {
	provider.mu.Lock()
	defer provider.mu.Unlock()

	if nil == provider.values {
		text, err := os.ReadFile(provider.File)
		if nil != err {
			// An absent file - or an absent directory - means "no secrets
			// here", exactly like FileProvider. Anything else (permission
			// denied, an unreadable mount) is a store that could not
			// answer, and swallowing it would fall through to a weaker
			// store.
			if !errors.Is(err, fs.ErrNotExist) && !errors.Is(err, syscall.ENOTDIR) {
				return nil, Fail("sekreto: dotenv provider cannot read " +
					provider.File + ": " + err.Error())
			}
			provider.values = map[string]string{}
		} else {
			provider.values = ParseDotenv(string(text))
		}
	}

	return provider.values, nil
}

func (provider *DotenvProvider) Lookup(name string) (string, bool, error) {
	key, err := EnvKey(name, provider.Prefix)
	if nil != err {
		return "", false, err
	}

	values, err := provider.load()
	if nil != err {
		return "", false, err
	}

	value, has := values[key]
	return value, has, nil
}

func (provider *DotenvProvider) Describe() string {
	return "dotenv:" + provider.File
}

// MemoryProvider holds literal values, keyed like environment variables.
// The spec uses this to test chain behaviour without touching the outside
// world.
type MemoryProvider struct {
	Values map[string]string
	Prefix string
}

func (provider *MemoryProvider) Lookup(name string) (string, bool, error) {
	key, err := EnvKey(name, provider.Prefix)
	if nil != err {
		return "", false, err
	}

	value, has := provider.Values[key]
	return value, has, nil
}

func (provider *MemoryProvider) Describe() string {
	if "" != provider.Prefix {
		return "memory:" + provider.Prefix
	}
	return "memory"
}

type FileProvider struct {
	Dir    string
	Prefix string
}

func (provider *FileProvider) Lookup(name string) (string, bool, error) {
	key, err := EnvKey(name, provider.Prefix)
	if nil != err {
		return "", false, err
	}

	file := filepath.Join(provider.Dir, key)

	raw, err := os.ReadFile(file)
	if nil != err {
		// An absent file - or an absent directory - means "no secrets
		// here", exactly like a missing .env. Anything else (permission
		// denied, an unreadable mount) is a store that could not answer.
		if errors.Is(err, fs.ErrNotExist) || errors.Is(err, syscall.ENOTDIR) {
			return "", false, nil
		}
		return "", false, Fail("sekreto: file provider cannot read " + file + ": " + err.Error())
	}

	text := string(raw)
	if strings.HasSuffix(text, "\n") {
		text = strings.TrimSuffix(strings.TrimSuffix(text, "\n"), "\r")
	}

	return text, true, nil
}

func (provider *FileProvider) Describe() string {
	return "file:" + provider.Dir
}

const ProviderExport = "provider"

const ErrorCode = "sekreto_error"

func ProviderPlugin(kind string, make func(spec *ProviderSpec) (Provider, error)) plugin.Definition {
	return plugin.Definition{
		Name: kind,
		Define: func(inst *plugin.Inst) error {
			spec, err := SpecOf(inst.Options())
			if nil != err {
				return err
			}

			provider, err := make(spec)
			if nil != err {
				var serr *SekretoError
				if errors.As(err, &serr) {
					return plugin.Fail(ErrorCode, serr.Message,
						map[string]any{"ref": inst.Ref(), "cause": serr.Message})
				}
				return err
			}

			inst.Export(ProviderExport, provider)
			return nil
		},
	}
}

// SpecOf reads a ProviderSpec back off a plugin instance's options map -
// the JSON shape OptionsOf produced, and the shape a config document
// would.
func SpecOf(options map[string]any) (*ProviderSpec, error) {
	// json.Marshal, not WriteJSON: these bytes are unmarshalled on the next
	// line and never leave the process, so HTML escaping cannot be observed
	// - Unmarshal reads \u003c back as <. WriteJSON is for what is emitted.
	text, err := json.Marshal(options)
	if nil != err {
		return nil, Fail("sekreto: unreadable provider options: " + err.Error())
	}

	spec := &ProviderSpec{}
	if err := json.Unmarshal(text, spec); nil != err {
		return nil, Fail("sekreto: unreadable provider options: " + err.Error())
	}

	return spec, nil
}

func OptionsOf(spec *ProviderSpec) (map[string]any, error) {
	text, err := json.Marshal(spec)
	if nil != err {
		return nil, Fail("sekreto: unwritable provider spec: " + err.Error())
	}

	options := map[string]any{}
	if err := json.Unmarshal(text, &options); nil != err {
		return nil, Fail("sekreto: unwritable provider spec: " + err.Error())
	}

	return options, nil
}

func Builtins() []plugin.Definition {
	return []plugin.Definition{
		ProviderPlugin("env", func(spec *ProviderSpec) (Provider, error) {
			return &EnvProvider{Prefix: spec.Prefix}, nil
		}),
		ProviderPlugin("memory", func(spec *ProviderSpec) (Provider, error) {
			values := spec.Values
			if nil == values {
				values = map[string]string{}
			}
			return &MemoryProvider{Values: values, Prefix: spec.Prefix}, nil
		}),
		ProviderPlugin("dotenv", func(spec *ProviderSpec) (Provider, error) {
			file := spec.File
			if "" == file {
				file = ".env"
			}
			return &DotenvProvider{File: file, Prefix: spec.Prefix}, nil
		}),
		ProviderPlugin("file", func(spec *ProviderSpec) (Provider, error) {
			return &FileProvider{Dir: spec.Dir, Prefix: spec.Prefix}, nil
		}),
	}
}

var Kinds = struct {
	Builtin []string
	Plugin  []string
}{
	Builtin: []string{"env", "memory", "dotenv", "file"},
	Plugin: []string{
		"hashicorp", "boru", "awssecrets", "awsparams", "gcpsecrets",
		"azuresecrets", "onepassword", "doppler", "infisical", "secretspec",
		"minivault",
	},
}

File: lua/src/sekreto/providers.lua
-- What a provider is, how a provider kind becomes a voxgig/plugin
-- definition - and the four BUILT-IN kinds.
--
-- A provider answers one question: "do you have this secret?" It returns
-- the value, or nil to mean "ask the next one". Nothing else about a
-- provider is visible to the caller - which is the point: an app reads
-- `api.token` and never learns whether it came from the environment, a
-- .env file, HashiCorp Vault or a boru vault.
--
-- Two failure shapes, and they are never interchangeable. A store that
-- does not hold the secret is a MISS (nil) - the chain carries on. A
-- store that could not answer - bad credentials, unreachable host,
-- missing configuration - is an ERROR: falling through there would
-- quietly reach for a weaker store.
--
-- A provider is a plain table with two functions, `lookup` and
-- `describe`, and no lifecycle. It is duck-typed: anything with a
-- callable `lookup` is a provider, whatever built it.
--
-- THIS MODULE REQUIRES NO SOCKET, NO DIGEST AND NO CHILD PROCESS. What
-- makes a kind built in is that it needs nothing of the platform beyond
-- reading a local file; every kind that opens a socket, signs a request
-- or spawns a process is a plugin under plugins/, its own module,
-- required only by a program that names it. `sekreto.plugins.net` - and
-- with it native/sekretonet.c, this port's only compiled artifact - is
-- reached by no file the core requires (docs/design/plugin-providers.md).
--
-- A port of typescript/src/provider/support.ts and
-- typescript/src/provider/builtin.ts, which are canonical.

local plugin = require('plugin')

local err = require('sekreto.err')
local name = require('sekreto.name')

local fail = err.fail

local M = {}

-- ------------------------------------------------------------ the specs

--- What a credential field reports about itself.
local function setornot(value)
  if nil == value or '' == value then
    return '[unset]'
  end
  return '[set]'
end

--- Printed without its credentials.
---
--- Lua's `tostring` on a bare table prints an address, but anything that
--- walks a spec - a config dumper, a logging helper - would otherwise
--- reach the Vault token, the AWS secret access key and the Azure client
--- secret. Fields holding a credential report whether they are set, never
--- what they are.
local SPECMT = {
  __tostring = function(spec)
    return 'ProviderSpec(kind=' .. tostring(spec.kind) ..
      ', name=' .. tostring(spec.name) ..
      ', addr=' .. tostring(spec.addr) ..
      ', token=' .. setornot(spec.token) ..
      ', secret=' .. setornot(spec.secret) ..
      ', clientsecret=' .. setornot(spec.clientsecret) ..
      ', auth=' .. tostring(spec.auth) .. ')'
  end,
  __name = 'ProviderSpec',
}

local AUTHMT = {
  __tostring = function(auth)
    return 'AuthSpec(method=' .. tostring(auth.method) ..
      ', mount=' .. tostring(auth.mount) ..
      ', role=' .. tostring(auth.role) ..
      ', jwtfile=' .. tostring(auth.jwtfile) ..
      ', roleid=' .. tostring(auth.roleid) ..
      ', jwt=' .. setornot(auth.jwt) ..
      ', secretid=' .. setornot(auth.secretid) .. ')'
  end,
  __name = 'AuthSpec',
}

--- The declarative form of a provider, as used in config and in the
--- shared spec. `kind` picks the provider; everything else is that
--- kind's own.
function M.spec(fields)
  return setmetatable(fields or {}, SPECMT)
end

--- Logging in to a vault instead of being handed a token. `method` is
--- `kubernetes` or `approle`; `mount` defaults to the method name.
function M.authspec(fields)
  return setmetatable(fields or {}, AUTHMT)
end

-- -------------------------------------------------------------- helpers

local function trimslash(text)
  if '/' == text:sub(-1) then
    return text:sub(1, #text - 1)
  end
  return text
end

M.trimslash = trimslash

local function nonempty(value)
  return nil ~= value and '' ~= value
end

M.nonempty = nonempty

--- The whole of a file, or nil plus the reason.
---
--- Absence - of the file OR of a directory on the way to it - is "no
--- secrets here" and answers nil with no reason, so the caller can tell
--- it from a permission error, which is a store that could not answer.
--- The reason string is not consulted; it comes from the C library's
--- strerror and follows the machine's locale.
local function readfile(path)
  local handle, why, code = io.open(path, 'rb')

  if nil == handle then
    -- ENOENT (2) and ENOTDIR (20): the file is not there, or a component
    -- of the path is not a directory. Both mean "no secrets here".
    if 2 == code or 20 == code then
      return nil, nil
    end
    return nil, ((why or 'cannot read'):gsub('^.-: ', ''))
  end

  local text = handle:read('a')
  handle:close()

  if nil == text then
    return nil, 'cannot read'
  end

  return text
end

M.readfile = readfile

-- ------------------------------------------------------- built-in kinds
--
-- The criterion for "built in" is needing nothing of the platform beyond
-- reading a local file: no socket, no TLS, no crypto, no child process.

--- Environment variables: `api.token` from `API_TOKEN`.
function M.env(prefix, source)
  return {
    lookup = function(secret)
      local key = name.envkey(secret, prefix)
      if nil == source then
        return os.getenv(key)
      end
      return source[key]
    end,
    describe = function()
      return 'env' .. (nonempty(prefix) and (':' .. prefix) or '')
    end,
  }
end

--- A `.env` file, read once, keyed exactly like the environment.
---
--- Loaded LAZILY. The `stores` corpus group puts a dotenv provider in a
--- chain and never looks anything up; an eager constructor would read
--- whatever `.env` happens to sit in the working directory.
function M.dotenv(file, prefix)
  local usefile = file or '.env'
  local values = nil

  local function load()
    if nil ~= values then
      return values
    end

    local text, why = readfile(usefile)

    if nil == text then
      if nil ~= why then
        fail('sekreto: dotenv provider cannot read ' .. usefile .. ': ' .. why)
      end
      -- An absent file, or an absent directory, means "no secrets here",
      -- exactly like the file provider.
      values = {}
    else
      values = name.parsedotenv(text)
    end

    return values
  end

  return {
    lookup = function(secret)
      return load()[name.envkey(secret, prefix)]
    end,
    describe = function()
      return 'dotenv:' .. usefile
    end,
  }
end

--- Literal values, keyed like environment variables. The shared spec uses
--- this to test chain behaviour without touching the outside world.
function M.memory(values, prefix)
  local usevalues = values or {}

  return {
    lookup = function(secret)
      return usevalues[name.envkey(secret, prefix)]
    end,
    describe = function()
      return 'memory' .. (nonempty(prefix) and (':' .. prefix) or '')
    end,
  }
end

--- A directory of one-secret-per-file entries, keyed like the
--- environment: `api.token` reads `<dir>/API_TOKEN`.
---
--- This is the shape of a mounted Kubernetes Secret, a Docker or Swarm
--- secret, and a systemd credentials directory, so those all work with no
--- further configuration. Read on every lookup, never cached. One
--- trailing newline is stripped - tools that write these files disagree
--- about it, and a newline is never part of a secret on purpose.
function M.file(dir, prefix)
  local usedir = dir or ''

  return {
    lookup = function(secret)
      local path = ('' == usedir) and name.envkey(secret, prefix)
        or (trimslash(usedir) .. '/' .. name.envkey(secret, prefix))

      local text, why = readfile(path)

      if nil == text then
        if nil ~= why then
          fail('sekreto: file provider cannot read ' .. path .. ': ' .. why)
        end
        return nil
      end

      if '\r\n' == text:sub(-2) then
        return text:sub(1, #text - 2)
      end
      if '\n' == text:sub(-1) then
        return text:sub(1, #text - 1)
      end

      return text
    end,
    describe = function()
      return 'file:' .. usedir
    end,
  }
end

-- ------------------------------ providers as voxgig/plugin definitions

--- The export key under which a provider definition publishes the
--- provider it built. `Sekreto` reads `<ref>/provider` off the host.
M.PROVIDER_EXPORT = 'provider'

--- The voxgig/plugin error code a SekretoError travels under when it is
--- raised inside a definition's `define`.
---
--- plugin wraps a code-less error raised by a callback as
--- `plugin_define_failed`, and keeps an error that already carries a
--- code. A provider that refuses its own configuration - `kv: 3`, a
--- missing project - raises a SekretoError, and that message is pinned by
--- the spec byte for byte, so it must come back out of the host exactly
--- as it went in. `providerplugin` gives it this code on the way in;
--- `Sekreto` turns it back into a SekretoError on the way out.
M.ERROR_CODE = 'sekreto_error'

--- A provider kind, as a voxgig/plugin definition.
---
--- This is the whole bridge between the two libraries. The definition's
--- `name` is the `kind` a spec names; its `define` reads the spec as
--- `inst:options()`, builds the provider with `make`, and exports it.
--- Nothing runs at `activate`: a provider opens nothing until its first
--- lookup, so there is nothing to capture - a provider that does hold a
--- resource acquires it there and lets the instance scope unwind it.
---
--- Every built-in and every plugin is made this way, so a custom provider
--- kind is one call:
---
---     providerplugin('mystore', function(spec) return mystore(spec.addr) end)
---
--- ONLY a SekretoError is re-raised with a code. Anything else a `make`
--- raises is not sekreto's to rewrite: it leaves this function as it
--- arrived, code-less, and the host wraps it as `plugin_define_failed`
--- naming the instance.
function M.providerplugin(kind, make)
  return {
    name = kind,
    define = function(inst)
      local ok, built = pcall(make, inst:options() or {})

      if not ok then
        if err.issekretoerror(built) then
          local message = err.message(built)
          plugin.types.fail(M.ERROR_CODE, message,
            plugin.types.map({ ref = inst.ref, cause = message }))
        end
        error(built, 0)
      end

      inst:export(M.PROVIDER_EXPORT, built)
    end,
  }
end

--- The four built-in provider kinds - the same four in every port. What
--- makes a kind built in is that it reads at most a local file: no
--- socket, no TLS, no crypto, no child process.
M.BUILTINS = {
  M.providerplugin('env', function(spec) return M.env(spec.prefix) end),
  M.providerplugin('memory', function(spec)
    return M.memory(spec.values or {}, spec.prefix)
  end),
  M.providerplugin('dotenv', function(spec)
    return M.dotenv(spec.file or '.env', spec.prefix)
  end),
  M.providerplugin('file', function(spec)
    return M.file(spec.dir or '', spec.prefix)
  end),
}

--- Every kind this library ships, built in or as a plugin, so that a kind
--- sekreto has never heard of can be told from one that was not passed
--- in.
M.KINDS = {
  builtin = { 'env', 'memory', 'dotenv', 'file' },
  plugin = {
    'hashicorp', 'boru', 'awssecrets', 'awsparams', 'gcpsecrets',
    'azuresecrets', 'onepassword', 'doppler', 'infisical', 'secretspec',
  },
}

return M
